openedx/openedx-platform
# Studio content api videos

## Description

https://2u-internal.atlassian.net/browse/TNL-10897

The purpose of this is to expose endpoints related to video uploads as public endpoint sas part of the studio content API MVP.

## Included refactorings

It was necessary to extract business logic from the video views for reuse to a service layer.
I did the same thing for transcripts, in preparation for the next upcoming PR that will expose transcript endpoints.

The files holding the service layer logic are cms/djangoapps/contentstore/transcript_storage_handlers.py and cms/djangoapps/contentstore/video_storage_handlers.py

## On the topic of unit tests

I did not include any unit tests for these reasons:
- We are working on an experimental MVP and these do not have priority. The new APIs are not yet enabled on prod but hidden behind a waffle flag. We do have some time constraints here.
- We can add tests in a later-stage PR before releasing the API MVP to prod.
- The actual functionality is all exhaustively tested. The untested code are only the urls.py and view files directly responsible for the new API. Endpoints have been manually tested to work, though.

## Out of scope

- Tests
- Swagger

## How to test

### Add an oauth-app to LMS django admin (instructions copied from[this opencraft article]( https://opencraft.com/use-the-open-edx-lms-for-authentication)):

Note: Everything we'll be doing is OAuth2, not OAuth1, which has been deprecated for quite some time.

We begin by registering a new application within the LMS. Log into [LMS admin panel](http://localhost:18000/) as the administrator. The default email and password for the admin on the devstack is edx@example.com and edx. From here, find the section with the header 'Django Oauth Toolkit.' Click 'Add' next to 'Applications':

![django-oauth-toolkit](https://github.com/openedx/edx-platform/assets/19345795/915d7632-9ab9-4443-9eff-b6d66331081b)

Every tool that needs to log into the LMS and perform actions on behalf of the user using OAuth has to be registered as an application. The form will initially populate with a unique client id and client secret. Save these and put them aside-- we'll need them later!

Leave 'User' blank, and set the Redirect URIs as below. Then, set the client type as 'Confidential' and the authorization grant type as 'Authorization Code'.

![oauth-example-app](https://github.com/openedx/edx-platform/assets/19345795/5bec06cf-2497-4a4f-9469-138bed8bf85c)


### Config

For this to work, you need to make a couple of changes in django-admin.
- To test video upload post endpoint without full AWS setup, add waffle flag "contentstore.mock_video_uploads" to studio django admin and set it to true. This will replace with some mock data.
- Waffle Flag `contentstore.enable_studio_content_api` must be set to true.
- Go to `<your_studio_base_url/admin/video_pipeline/videouploadsenabledbydefault/add/`, check the `enabled` checkbox, and save.
- Waffle Flag `contentstore.video_upload_pipeline` should be set to **False**.

### Testing requests with Insomnia

- Create an insomnia environment with this value, and fill in the client id and secret you copied from the django-admin oauth app:
```json
{
	"clientId": <your clientId>,
	"clientSecret": <your clientSecret
}
```
- Import this course to studio for testing:
[course.czpej34u.tar.gz](https://github.com/openedx/edx-platform/files/12174667/course.czpej34u.tar.gz)
- Go to insomnia, click the top breadcrumbs at the top center (Should say something like "Insomnia / My Collection" - click on Insomnia, which should bring you to a home page where you can create a new collection. Create a new collection.
- Again, on the breadcrumbs, click on the dropdown and select import
- Unzip and Import this file to your collection:
[Insomnia_video_api.json.zip](https://github.com/openedx/edx-platform/files/12174695/Insomnia_video_api.json.zip)
- Now you should have the relevant requests.
- You need to replace `course-v1:MITx+aga+adg` with your course ID for every request.
- To test the `videos/features` endpoint, you can add a new Waffle **Switch** `videos.video_image_upload_enabled`, which should switch the return value of the corresponding JSON key for the endpoint; and add a Waffle **Flag** `video_config.public_video_share`.
